blog/tidy-market-microstructure/index.qmd

```diff
@@ -0,0 +1,1674 @@
+  summarise(across(contains("quoted"), 
+                   ~round(mean(.), digits = 2))) |>
+  pivot_longer(-ticker) |>
+  as_tibble() |>
+  pivot_wider(names_from = name, values_from = value)
+
+```
+:::
+
+## Consolidated liquidity in fragmented markets
+With the competition between exchanges, liquidity is dispersed across venues. For example, if there is a change to the market structure at the LSE, it is typically not sufficient to analyze liquidity at LSE alone. If liquidity is reduced at the LSE, it may simultaneously be boosted at the TQE. To assess the overall market quality, which may be most relevant for welfare, it is often necessary to consider the *consolidated* liquidity. 
+
+In Europe, the consolidated liquidity is sometimes referred to as the European Best Bid and Offer (EBBO). The terminology follows in the footsteps of the US market, where the *National* Best Bid and Offer (NBBO) is transmitted to the market on continuous basis. To obtain the EBBO, one needs to merge the LOB data from each relevant venue, and then determine the EBBO prices and depths. In the code below, we show step-by-step how to do that.
+
+**Retaining only the last quote update in each interval.**
+Quote updates tend to cluster and it is common that several observations have identical timestamps. Multiple observations at one timestamp can be due to several investors responding to the same events, or that one market order leads to several LOB updates as it is executed against multiple limit orders. When matching quotes across venues, we need to restrict the number of observations per unit of time to one. There is no sensible way to distinguish observations with identical timestamps. In lack of a better approach, we retain the last observation in each interval. 
+
+::: {.panel-tabset group="language"}
+## data.table
+```{r Last entry per unit of time}
+# Retain only the last observation per unit of time
+# The function `duplicated` returns `TRUE` if the observation is a duplicate of another 
+# observation based on the columns given in the `by` option, and `FALSE` otherwise.
+# The option `fromLast = TRUE` ensures that the last rather than the first observation 
+# in each millisecond that returns `FALSE`.
+quotes <- quotes[!duplicated(quotes, fromLast = TRUE, by = c("ticker", "date", "time"))]
+```
+## tidyverse (using dtplyr)
+```{r}
+tv_quotes <- tv_quotes |> 
+  group_by(ticker, date, time) |>
+  slice(n()) |> 
+  ungroup() |>
+  lazy_dt()
+```
+:::
+
+**Merging quotes from different venues.**
+We are now ready to match the quotes from the two exchanges. First, we create separate quote data sets for the two exchanges. Second, we merge the two by matching on date and time. Third, we forward-fill quotes from both venues, such that for each LSE quote we know the prevailing TQE quote, and vice versa. The validity of this is ensured by the option `sort = TRUE` in the `merge` function, which returns a data.table that is sorted on the matching variables.
```

Comment:
"The validity of this is ensured by the option `sort = TRUE` in the `merge` function, which returns a data.table that is sorted on the matching variables." -> no equivalent statement about dtplyr